biogeophys/FatesHydroWTFMod.F90

```diff
@@ -0,0 +1,1136 @@
+     real(r8) :: th_res   ! Residual volumentric water content          [m3/m3]
+     real(r8) :: pinot    ! osmotic potential at full turger            [MPa]
+     real(r8) :: epsil    ! bulk elastic modulus                        [MPa]
+     real(r8) :: rwc_fd   ! total RWC @ which elastic drainage begins     [-]
+     real(r8) :: cap_corr ! correction for nonzero psi0x
+     real(r8) :: cap_int  ! intercept of capillary region of curve
+     real(r8) :: cap_slp  ! slope of capillary region of curve
+     integer  :: pmedia   ! self describing porous media index
+
+     real(r8) :: psi_max     ! psi matching max_sf_interp where we start linear interp
+     real(r8) :: dpsidth_max ! dpsi_dth where we start linear interp
+     real(r8) :: psi_min     ! psi matching min_sf_interp
+     real(r8) :: dpsidth_min ! dpsi_dth where we start min interp
+
+   contains
+     procedure :: th_from_psi     => th_from_psi_tfs
+     procedure :: psi_from_th     => psi_from_th_tfs
+     procedure :: dpsidth_from_th => dpsidth_from_th_tfs
+     procedure :: set_wrf_param   => set_wrf_param_tfs
+     procedure :: bisect_pv
+  end type wrf_type_tfs
+
+  ! Water Conductivity Function
+  type, public, extends(wkf_type) :: wkf_type_tfs
+     real(r8) :: p50          ! matric potential at 50% conductivity loss   [Mpa]
+     real(r8) :: avuln        ! vulnerability curve parameter
+     real(r8) :: th_sat       ! volumetric water content at saturation
+
+   contains
+     procedure :: ftc_from_psi      => ftc_from_psi_tfs
+     procedure :: dftcdpsi_from_psi => dftcdpsi_from_psi_tfs
+     procedure :: set_wkf_param     => set_wkf_param_tfs
+  end type wkf_type_tfs
+
+
+contains
+
+  ! =====================================================================================
+  ! Functional definitions follow here
+  ! Start off by writing the base types, which ultimately should never be pointed to.
+  ! =====================================================================================
+
+  subroutine set_wrf_param_base(this,params_in)
+    class(wrf_type)     :: this
+    real(r8),intent(in) :: params_in(:)
+    write(fates_log(),*) 'The base water retention function'
+    write(fates_log(),*) 'should never be actualized'
+    write(fates_log(),*) 'check how the class pointer was setup'
+    call endrun(msg=errMsg(sourcefile, __LINE__))
+  end subroutine set_wrf_param_base
+  subroutine set_wkf_param_base(this,params_in)
+    class(wkf_type)     :: this
+    real(r8),intent(in) :: params_in(:)
+    write(fates_log(),*) 'The base water retention function'
+    write(fates_log(),*) 'should never be actualized'
+    write(fates_log(),*) 'check how the class pointer was setup'
+    call endrun(msg=errMsg(sourcefile, __LINE__))
+  end subroutine set_wkf_param_base
+  function th_from_psi_base(this,psi) result(th)
+    class(wrf_type)     :: this
+    real(r8),intent(in) :: psi
+    real(r8)            :: th
```

Comment:
It is implicitly intent out because it is defined in the result() operator.  I suppose we could label this as intent out to re-inforce the point. I will see if the compiler has issues with this.

Reply:
I did check this and at least the gnu compilers don't allow us to specify the intent of the result, as the result of a function is implied to be out.